More warnings with CUDA 11.8 and GCC 11.3.0

I'm getting a couple of warnings on the current `develop` branch:

```
/build/ginkgo-hpc-git/src/ginkgo/cuda/base/pointer_mode_guard.hpp: In destructor ‘gko::kernels::cuda::cublas::pointer_mode_guard::~pointer_mode_guard()’:
/build/ginkgo-hpc-git/src/ginkgo/cuda/base/pointer_mode_guard.hpp:82:10: warning: ‘bool std::uncaught_exception()’ is deprecated: use 'std::uncaught_exceptions()' instead [-Wdeprecated-declarations]
   82 |         if (std::uncaught_exception()) {
      |          ^~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/11.3.0/include/c++/exception:109:1: note: declared here
  109 |   bool uncaught_exception() _GLIBCXX_USE_NOEXCEPT __attribute__ ((__pure__));
      | ^ ~~~~~~~~~~~~~~~~
```
(the same appears on a few more places in `pointer_mode_guard.hpp`)

```
/build/ginkgo-hpc-git/src/ginkgo/cuda/test/base/index_set.cu(84): warning #177-D: function "<unnamed>::index_set::assert_equal_arrays" was declared but never referenced
```

From doxygen 1.9.5:
```
warning: Tag 'CLASS_DIAGRAMS' at line 85 of file '/build/ginkgo-hpc-git/src/build/doc/Doxyfile-usr' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
```

At runtime, some tests failed:
```
100/300 Test  #91: omp/test/reorder/rcm_kernels .............................Subprocess aborted***Exception:   0.18 sec
Running main() from /build/gtest/src/googletest-release-1.12.1/googletest/src/gtest_main.cc
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from Rcm
[ RUN      ] Rcm.OmpPermutationIsRcmOrdered
/usr/include/c++/12.2.0/bits/stl_vector.h:1123: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = int; _Alloc = gko::ExecutorAllocator<int>; reference = int&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.

223/300 Test #226: test/factorization/par_ilu_kernels_omp ...................Subprocess aborted***Exception:   0.26 sec
Running main() from /build/gtest/src/googletest-release-1.12.1/googletest/src/gtest_main.cc
[==========] Running 56 tests from 8 test suites.
[----------] Global test environment set-up.
[----------] 7 tests from ParIlu/<float, int>, where TypeParam = std::tuple<float, int>
[ RUN      ] ParIlu/<float, int>.KernelAddDiagonalElementsSortedEquivalentToRef
[       OK ] ParIlu/<float, int>.KernelAddDiagonalElementsSortedEquivalentToRef (85 ms)
[ RUN      ] ParIlu/<float, int>.KernelAddDiagonalElementsUnsortedEquivalentToRef
/usr/include/c++/12.2.0/bits/uniform_int_dist.h:97: std::uniform_int_distribution<_IntType>::param_type::param_type(_IntType, _IntType) [with _IntType = int]: Assertion '_M_a <= _M_b' failed.

        Start 227: test/factorization/par_ilu_kernels_cuda
224/300 Test #227: test/factorization/par_ilu_kernels_cuda ..................Subprocess aborted***Exception:   2.55 sec
Running main() from /build/gtest/src/googletest-release-1.12.1/googletest/src/gtest_main.cc
[==========] Running 56 tests from 8 test suites.
[----------] Global test environment set-up.
[----------] 7 tests from ParIlu/<float, int>, where TypeParam = std::tuple<float, int>
[ RUN      ] ParIlu/<float, int>.KernelAddDiagonalElementsSortedEquivalentToRef
[       OK ] ParIlu/<float, int>.KernelAddDiagonalElementsSortedEquivalentToRef (2205 ms)
[ RUN      ] ParIlu/<float, int>.KernelAddDiagonalElementsUnsortedEquivalentToRef
/usr/include/c++/12.2.0/bits/uniform_int_dist.h:97: std::uniform_int_distribution<_IntType>::param_type::param_type(_IntType, _IntType) [with _IntType = int]: Assertion '_M_a <= _M_b' failed.

245/300 Test #248: test/matrix/matrix_omp ...................................Subprocess aborted***Exception:   0.16 sec
Running main() from /build/gtest/src/googletest-release-1.12.1/googletest/src/gtest_main.cc
[==========] Running 323 tests from 17 test suites.
[----------] Global test environment set-up.
[----------] 19 tests from Matrix/DenseWithDefaultStride, where TypeParam = DenseWithDefaultStride
[ RUN      ] Matrix/DenseWithDefaultStride.SpMVIsEquivalentToRef
/usr/include/c++/12.2.0/bits/uniform_int_dist.h:97: std::uniform_int_distribution<_IntType>::param_type::param_type(_IntType, _IntType) [with _IntType = long int]: Assertion '_M_a <= _M_b' failed.

        Start 249: test/matrix/matrix_cuda
246/300 Test #249: test/matrix/matrix_cuda ..................................Subprocess aborted***Exception:   3.08 sec
Running main() from /build/gtest/src/googletest-release-1.12.1/googletest/src/gtest_main.cc
[==========] Running 399 tests from 21 test suites.
[----------] Global test environment set-up.
[----------] 19 tests from Matrix/DenseWithDefaultStride, where TypeParam = DenseWithDefaultStride
[ RUN      ] Matrix/DenseWithDefaultStride.SpMVIsEquivalentToRef
/usr/include/c++/12.2.0/bits/uniform_int_dist.h:97: std::uniform_int_distribution<_IntType>::param_type::param_type(_IntType, _IntType) [with _IntType = long int]: Assertion '_M_a <= _M_b' failed.
```

--- a/cuda/test/base/CMakeLists.txt
+++ b/cuda/test/base/CMakeLists.txt
@@ -1,6 +1,6 @@
 ginkgo_create_cuda_test(array)
 ginkgo_create_cuda_test(cuda_executor)
-ginkgo_create_cuda_test(index_set)
+ginkgo_create_test(index_set)
 ginkgo_create_test(cuda_executor_reset ADDITIONAL_LIBRARIES Threads::Threads)
 if(GINKGO_HAVE_HWLOC)
     find_package(NUMA REQUIRED)

--- a/cuda/test/base/index_set.cpp
+++ b/cuda/test/base/index_set.cpp
@@ -81,15 +81,6 @@ protected:
         }
     }
 
-    static void assert_equal_arrays(const T num_elems, const T* a, const T* b)
-    {
-        if (num_elems > 0) {
-            for (auto i = 0; i < num_elems; ++i) {
-                EXPECT_EQ(a[i], b[i]);
-            }
-        }
-    }
-
     std::shared_ptr<const gko::Executor> exec;
     std::shared_ptr<gko::CudaExecutor> cuda;
 };

--- a/doc/conf/Doxyfile-pdf.in
+++ b/doc/conf/Doxyfile-pdf.in
@@ -24,7 +24,6 @@ LATEX_TIMESTAMP        = NO
 PREDEFINED = @default_predefined_macros@ protected=private
 
 # hide all unnecessary graphs
-CLASS_DIAGRAMS         = NO
 CLASS_GRAPH            = NO
 COLLABORATION_GRAPH    = NO
 GROUP_GRAPHS           = NO

--- a/doc/conf/Doxyfile.in
+++ b/doc/conf/Doxyfile.in
@@ -82,8 +82,7 @@ ALLEXTERNALS           = NO
 EXTERNAL_GROUPS        = YES
 
 # Graph generation
-CLASS_DIAGRAMS         = YES
-CLASS_GRAPH            = NO
+CLASS_GRAPH            = YES
 COLLABORATION_GRAPH    = YES
 GROUP_GRAPHS           = YES
 GRAPHICAL_HIERARCHY    = YES

--- a/hip/test/base/CMakeLists.txt
+++ b/hip/test/base/CMakeLists.txt
@@ -1,5 +1,5 @@
 ginkgo_create_hip_test(hip_executor)
-ginkgo_create_hip_test(index_set)
+ginkgo_create_test(index_set)
 ginkgo_create_test(hip_executor_reset ADDITIONAL_LIBRARIES Threads::Threads)
 if(GINKGO_HAVE_HWLOC)
     find_package(NUMA REQUIRED)

--- a/hip/test/base/index_set.cpp
+++ b/hip/test/base/index_set.cpp
@@ -81,15 +81,6 @@ protected:
         }
     }
 
-    static void assert_equal_arrays(const T num_elems, const T* a, const T* b)
-    {
-        if (num_elems > 0) {
-            for (auto i = 0; i < num_elems; ++i) {
-                EXPECT_EQ(a[i], b[i]);
-            }
-        }
-    }
-
     std::shared_ptr<const gko::Executor> exec;
     std::shared_ptr<gko::HipExecutor> hip;
 };

